cpp/src/arrow/testing/random.cc

```diff
@@ -262,5 +262,62 @@ std::shared_ptr<Array> RandomArrayGenerator::Offsets(int64_t size, int32_t first
   return std::make_shared<Int32Array>(array_data);
 }
 
+struct RandomArrayGeneratorOfImpl {
+  Status Visit(const NullType&) {
+    DCHECK_NE(null_probability_, 0.0);
```

Comment:
This check can perhaps be dropped, I could see this being an issue if you were running an array of null probabilities across some type parameters

Reply:
alright